arangod/Pregel/Actor/Actor.h

```diff
@@ -0,0 +1,83 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2022 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Markus Pfeiffer
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <Inspection/VPackWithErrorT.h>
+#include <memory>
+
+#include "MPSCQueue.h"
+
+namespace arangodb::pregel::actor {
+
+template<typename Scheduler, typename MessageHandler, typename State,
+         typename Message>
+struct Actor {
+  Actor(Scheduler& schedule, State initialState)
+      : schedule(schedule), state(initialState) {}
+
+  Actor(Scheduler& schedule, State initialState, std::size_t batchSize)
+      : batchSize(batchSize), schedule(schedule), state(initialState) {}
+
+  void process(std::unique_ptr<Message> msg) {
+    inbox.push(std::move(msg));
+    kick();
```

Comment:
You can use mpsc stack to understand need kick or not.

Check async mutex or strand implemention: https://github.com/YACLib/YACLib/blob/main/src/exe/strand.cpp#L29

Reply:
I don't understand! When we push a message, we always need to call `kick` to schedule a task to ensure that it will be processed. Not sure what an async mutex or strand has to do with it. Also, why do you think `kick` requires a global mutex?

Reply:
Because kick push task to executor, which commonly shared between threads.

Not you don't need kick every push you have cycle in work, while you're in cycle you don't need to call kick.
Question: how to synchronize kick with end of cycle? One of answers: strand code

Reply:
```auto* expected = _jobs.load(std::memory_order_relaxed);
  do {
    job.next = expected == Mark() ? nullptr : expected;
  } while (!_jobs.compare_exchange_weak(expected, &job));
  if (expected == Mark()) {
   kick();
  }

```

Reply:
Also current code contains bug, what if I make two pushes, from different threads:

Them doing something like

Thread1:
Add task1
kick
pop task1
thinks what's all
Thread2:
push task2
kick
thinks other thread execute it (cas on busy)
Thread1
store false to the busy
end

Reply:
True, we do not need to schedule a new task if one is already scheduled or being executed. Yes, this matches the execution model of a strand, but we do not have something like that in our code right now, and it is easy enough to simulate. I haven't checked the code yet, so it is possible there are still some bugs.
FWIW - this is just a poor mans implementation of an actor model by using our current scheduler (which is not really designed for this). The main goal is to gather experience, so this is not meant to be tuned for performance.

Reply:
Your message for me doesn't matches with implementation Vyukov queue.